### PR TITLE
feat(audio): Add lazy-generated TTS briefing audio

### PR DESCRIPTION
**Pull request by allenhutchison**
## Summary

Transforms the daily briefing from email-only into an auto-generated podcast. Users click **Play** in the feed UI or email link; audio is generated on-demand via `gemini-3.1-flash-tts-preview`, cached as a DB blob, and streamed on subsequent requests.

### Pipeline
```
Briefing text (existing) → Audio script rewrite (Gemini Flash) → TTS render (gemini-3.1-flash-tts-preview) → PCM → MP3 (ffmpeg) → DB blob
```

## Design decisions

- **Lazy generation**: Audio is only generated when a user first clicks Play, not when the text briefing is created. Cost scales with actual usage, not user count.
- **Claim/dedup pattern**: Mirrors the existing `claim_briefing_generation` concurrency model — concurrent Play clicks won't double-generate.
- **DB blob storage**: Audio stored as `LargeBinary` in `daily_briefings` table. Avoids filesystem dependency (the podcast download dir is `/tmp` and not persisted). PostgreSQL TOAST handles large values out-of-line automatically.
- **Deferred blob loading**: Feed reads use `defer(DailyBriefing.audio_data)` to avoid pulling MB-scale blobs on normal list requests.
- **Audio script rewrite**: The newsletter-style briefing is restructured for spoken delivery — top stories segment, quick rundown of remaining episodes, reflective sign-off with connection insight.
- **HTTP Range support**: The GET endpoint supports `Range` headers for audio seeking in the browser `<audio>` player.

## Changes

| Area | Files | What |
|------|-------|------|
| Config | `src/config.py` | `GEMINI_TTS_MODEL`, `GEMINI_TTS_VOICE`, `BRIEFING_AUDIO_SAMPLE_RATE`, `BRIEFING_AUDIO_RETENTION_DAYS` |
| Model | `src/db/models.py` | 5 new columns on `DailyBriefing`: `audio_data`, `audio_mime_type`, `audio_status`, `audio_duration_sec`, `audio_generated_at` |
| Migration | `alembic/versions/...` | Adds nullable audio columns |
| Repository | `src/db/repository.py` | 5 new methods: `get_briefing_by_id`, `claim_briefing_audio`, `update_briefing_audio_status`, `update_briefing_audio`, `clear_audio_data_before` + deferred blob loading in `get_daily_briefings_in_range` |
| Script gen | `src/services/briefing_generator.py` | `generate_audio_script()` — rewrites briefing as spoken prose with audio tags |
| TTS | `src/services/tts.py` (new) | `render_tts_to_mp3()` — Gemini TTS API → base64 PCM → ffmpeg → MP3 bytes |
| Orchestration | `src/services/briefing_audio.py` (new) | `generate_briefing_audio()` + `get_audio_url_or_trigger()` with claim/dedup |
| API | `src/web/app.py` | `POST /api/feed/briefing/{id}/audio` (trigger), `GET .../audio` (stream with Range) |
| Feed response | `src/services/feed_service.py` | Adds `audio_status`, `audio_url`, `audio_duration_sec` to briefing response |
| Email | `src/services/email_renderer.py` | "▶ Listen to this briefing" link in email HTML |
| Email worker | `src/workflow/workers/email_digest.py` | Passes briefing `id` through to renderer for audio link |
| Frontend | `src/web/static/feed.html` | `<audio>` player + Generate button + `?play=` param handler for email links |
| Tests | 4 new test files | 30 tests covering TTS, audio script, orchestration, repo methods |

## Token budget check

For a typical 13-episode briefing (~2,500 chars text + highlights):
- **Audio script input**: ~2,500 tokens (well under 8,192 limit)
- **TTS input**: ~2,500 tokens (the script, not raw transcripts)
- **TTS output**: ~5,000 audio tokens for 3-4 min of speech (well under 16,384 limit)
- No chunking or stitching needed.

## Cost model

| Scenario | Per briefing | 100 users/day |
|----------|-------------|---------------|
| Script rewrite (Gemini Flash, ~3K tokens) | ~\$0.001 | \$0.10/day |
| TTS generation (~5K output tokens) | ~\$0.05 | \$5.00/day |

With lazy generation at 10% adoption: **~\$0.50/day**.

## Test plan

- [x] All 1147 existing tests pass (no regressions)
- [x] 30 new tests pass (TTS, audio script, orchestration, repo)
- [x] `ruff check` passes on all new files
- [ ] Manual: run `doppler run -- alembic upgrade head` on dev DB
- [ ] Manual: start server, load feed, click Generate audio, verify MP3 plays
- [ ] Manual: verify email "Listen" link opens feed page and triggers audio
- [ ] Manual: verify audio seeking works (Range requests)

## Deployment notes

No new Doppler secrets required (uses existing `GEMINI_API_KEY`). New env vars have defaults:
```
GEMINI_TTS_MODEL=gemini-3.1-flash-tts-preview
GEMINI_TTS_VOICE=Puck
BRIEFING_AUDIO_SAMPLE_RATE=24000
BRIEFING_AUDIO_RETENTION_DAYS=30
```

ffmpeg/ffprobe already required by the project for transcription.

🤖 Generated with opencode

## Summary by CodeRabbit

* **New Features**
  * Added lazy daily briefing audio generation with cached `pending/ready/failed` states, safe concurrent claiming, and MP3 streaming.
  * Added feed UI and API support to trigger audio generation and stream audio with HTTP byte-range support.
  * Added spoken audio script generation and automatic audio retention cleanup.
  * Email digests now include a “Listen to this briefing” link when available.
* **Configuration**
  * Added environment-driven TTS model/voice and audio sample rate/retention settings with validation.
* **Bug Fixes**
  * Improved timezone fallback to use `datetime.UTC`.
* **Tests**
  * Added coverage for audio orchestration, repository behavior, script/TTS generation, and API + Range handling.

### CODE REVIEW

File: src/config.py [hunk 1]
```diff
@@ -70,6 +70,18 @@ def __init__(self, env_file=None):
...
+        self.GEMINI_TTS_MODEL = os.getenv(
+            "GEMINI_TTS_MODEL", "gemini-3.1-flash-tts-preview"
+        )
+        self.GEMINI_TTS_VOICE = os.getenv("GEMINI_TTS_VOICE", "Puck")
+        self.BRIEFING_AUDIO_SAMPLE_RATE = int(
+            os.getenv("BRIEFING_AUDIO_SAMPLE_RATE", "24000")
+        )
+        self.BRIEFING_AUDIO_RETENTION_DAYS = int(
+            os.getenv("BRIEFING_AUDIO_RETENTION_DAYS", "30")
+        )
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/config.py:87]: Fixed in b210389 — added range validation at config load time with ValueError for invalid values.

File: src/db/repository.py [hunk 2]
```diff
@@ -3671,6 +3719,88 @@ def release_briefing_claim(self, user_id: str, briefing_date: date) -> None:
...
+                    DailyBriefing.id == briefing_id,
+                    or_(
+                        DailyBriefing.audio_status.is_(None),
+                        DailyBriefing.audio_status == "failed",
+                    ),
+                )
+                .values(audio_status="generating")
+            )
+            session.commit()
+            return result.rowcount > 0
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/db/repository.py]: Fixed in b210389 — added `audio_claimed_at` column and modified `claim_briefing_audio` to allow reclaiming stale "generating" claims older than 30 minutes (configurable via `stale_claim_timeout_sec`).

File: src/services/briefing_audio.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # Another request is generating — tell client to poll
+    if briefing.audio_status == "generating":
+        return BRIEFING_AUDIO_PENDING
+
+    # Claim the slot atomically
+    claimed = repository.claim_briefing_audio(briefing_id)
+    if not claimed:
+        # Someone beat us to it
+        return BRIEFING_AUDIO_PENDING
+
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/services/briefing_audio.py]: Fixed in b210389 — both `generate_briefing_audio` and `get_audio_url_or_trigger` now detect the inconsistent "ready"+no-data state and reset to "failed" before claiming, allowing regeneration. Added a regression test (`test_ready_without_data_recovers`).

File: src/services/briefing_generator.py [hunk 4]
```diff
@@ -367,3 +367,70 @@ def generate_digest_briefing(episodes: list, config: Config) -> dict | None:
...
+        text = response.text.strip() if response.text else ""
+        if text:
+            return text
+
+        logger.warning("Empty response from Gemini for audio script")
+        return None
+
+    except Exception:
+        logger.exception("Failed to generate audio script")
+        return None
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/services/briefing_generator.py:444]: Fixed in b210389 — `generate_audio_script` now validates input with `DigestBriefing.model_validate()` and serializes `model_dump()` instead of passing the raw dict.

File: src/services/tts.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                        prebuilt_voice_config=types.PrebuiltVoiceConfig(
+                            voice_name=config.GEMINI_TTS_VOICE,
+                        )
+                    )
+                ),
+            ),
+        )
+    except Exception:
+        logger.exception("Gemini TTS API call failed")
+        return None, None
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/services/tts.py:94]: Fixed in b210389 — added `_retry_tts_call` with exponential backoff (3 retries, 429/5xx detection), mirroring the existing `_retry_generate_content` pattern in briefing_generator.py.

File: src/services/tts.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                "-ac", "1",
+                "-i", "pipe:0",
+                "-codec:a", "libmp3lame",
+                "-b:a", "64k",
+                "-f", "mp3",
+                "pipe:1",
+            ],
+            input=pcm_bytes,
+            capture_output=True,
+        )
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/services/tts.py:139]: Fixed in b210389 — added `timeout=120` to both ffmpeg and ffprobe subprocess calls, with explicit `subprocess.TimeoutExpired` handling.

File: src/web/app.py [hunk 7]
```diff
@@ -1146,6 +1147,109 @@ async def generate_feed_briefing(
...
+def _parse_range(range_header: str, total: int) -> tuple[int, int]:
+    """Parse a Range header value into (start, end) inclusive."""
+    try:
+        unit, ranges = range_header.split("=")
+        start_str, end_str = ranges.split("-")
+        start = int(start_str) if start_str else 0
+        end = int(end_str) if end_str else total - 1
+        return start, min(end, total - 1)
+    except Exception:
+        return 0, total - 1
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/web/app.py]: Fixed in b210389 — `_parse_range` now returns `None` for malformed/unsatisfiable ranges (invalid unit, start > end, out of bounds). The endpoint returns 416 with `Content-Range: bytes */{total}` when parsing fails.

File: src/web/static/feed.html [hunk 8]
```diff
@@ -239,11 +241,61 @@ <h3 class="text-lg font-bold text-gray-900">${escapeHtml(briefing.headline)}</h3
...
+        }
+
+        async function triggerBriefingAudio(briefingId, btn) {
+            if (btn) {
+                btn.disabled = true;
+                btn.textContent = 'Generating audio...';
+            }
+            try {
+                const res = await fetch(`/api/feed/briefing/${briefingId}/audio`, { method: 'POST' });
+                const data = await res.json();
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/web/static/feed.html]: Fixed in b210389 — added `credentials: 'include'` to the fetch call.

File: src/web/static/feed.html [hunk 9]
```diff
@@ -239,11 +241,61 @@ <h3 class="text-lg font-bold text-gray-900">${escapeHtml(briefing.headline)}</h3
...
+                    const audio = document.createElement('audio');
+                    audio.controls = true;
+                    audio.autoplay = true;
+                    audio.src = data.audio_url;
+                    audio.className = 'w-full mt-3';
+                    btn.replaceWith(audio);
+                } else if (data.status === 'pending') {
+                    if (btn) btn.textContent = 'Generating...';
+                    setTimeout(() => triggerBriefingAudio(briefingId, btn), 5000);
+                } else {
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/web/static/feed.html:297]: Fixed in b210389 — added `attempt` parameter with `MAX_ATTEMPTS=12` (~60s at 5s intervals). On timeout, resets the button and shows an alert.

File: src/web/static/feed.html [hunk 10]
```diff
@@ -566,7 +618,27 @@ <h3 class="font-semibold text-gray-900 leading-snug">${escapeHtml(ep.title)}</h3
...
+            if (!playId) return;
+            // Wait for the briefing card to appear, then trigger audio
+            const maxWait = 10000;
+            const start = Date.now();
+            const tryTrigger = () => {
+                const safePlayId = CSS.escape(playId);
+                const btn = document.querySelector(`button[onclick*="${safePlayId}"]`);
+                if (btn) {
+                    btn.click();
+                } else if (Date.now() - start < maxWait) {
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/web/static/feed.html:658]: Fixed in b210389 — `handlePlayParam` now checks for an `<audio>` element first (already-ready state) and calls `.play()` on it, falling back to the generate button if not found.

File: src/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/config.py:87]: Also addressed in b210389 — replaced the hardcoded domain with `config.APP_BASE_URL` and URL-encoded the briefing_id with `urllib.parse.quote`.

File: src/services/briefing_audio.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/services/briefing_audio.py]: Regression test added in b210389 — `test_ready_without_data_recovers` verifies the recovery path when audio_status='ready' but audio_data is None.

**Comment by allenhutchison** [comment]
## Round 2 Summary

All 10 actionable comments + 2 nitpicks addressed in b210389.

| Comment | File | Fix |
|---------|------|-----|
| Validate config bounds | `src/config.py` | Range validation with ValueError at startup |
| Stale-claim recovery | `src/db/repository.py` | `audio_claimed_at` column + 30min timeout reclaim |
| Ready-without-data recovery | `src/services/briefing_audio.py` | Reset to "failed" before claiming |
| Validate briefing_data | `src/services/briefing_generator.py` | `DigestBriefing.model_validate()` before serialization |
| TTS retry with backoff | `src/services/tts.py` | `_retry_tts_call` with exponential backoff |
| Subprocess timeouts | `src/services/tts.py` | 120s timeout + `TimeoutExpired` handling |
| Malformed Range headers | `src/web/app.py` | `_parse_range` returns None → 416 response |
| Missing credentials | `src/web/static/feed.html` | `credentials: 'include'` on fetch |
| Infinite polling | `src/web/static/feed.html` | MAX_ATTEMPTS=12 (~60s) with user-facing error |
| Deep-link audio fallback | `src/web/static/feed.html` | Check `<audio>` element before button |
| Hardcoded email URL (nit) | `src/services/email_renderer.py` | `config.APP_BASE_URL` + `urllib.parse.quote` |
| Regression test (nit) | `tests/test_briefing_audio.py` | `test_ready_without_data_recovers` added |

All 1148 tests pass. Waiting for re-review.

**Comment by allenhutchison** [comment]
## Round 2 Summary

All 10 actionable comments + 2 nitpicks addressed in b210389.

| Comment | File | Fix |
|---------|------|-----|
| Validate config bounds | config.py | Range validation with ValueError at startup |
| Stale-claim recovery | repository.py | audio_claimed_at column + 30min timeout reclaim |
| Ready-without-data recovery | briefing_audio.py | Reset to failed before claiming |
| Validate briefing_data | briefing_generator.py | DigestBriefing.model_validate() before serialization |
| TTS retry with backoff | tts.py | _retry_tts_call with exponential backoff |
| Subprocess timeouts | tts.py | 120s timeout + TimeoutExpired handling |
| Malformed Range headers | app.py | _parse_range returns None -> 416 response |
| Missing credentials | feed.html | credentials: include on fetch |
| Infinite polling | feed.html | MAX_ATTEMPTS=12 (~60s) with user-facing error |
| Deep-link audio fallback | feed.html | Check audio element before button |
| Hardcoded email URL (nit) | email_renderer.py | config.APP_BASE_URL + urllib.parse.quote |
| Regression test (nit) | test_briefing_audio.py | test_ready_without_data_recovers added |

All 1148 tests pass. Waiting for re-review.

**Comment by allenhutchison** [comment]
## Round 3 — outside-diff-range comment

Fixed in c08ece1 — replaced the broad `except Exception` in `generate_audio_script` with specific exception types (`ValidationError`, `ValueError`, `TypeError`) for input/API errors, keeping a broad `except Exception` as a last resort with a more descriptive log message.

All 1148 tests pass.

**Comment by allenhutchison** [comment]
## PR Review — Lazy-generated TTS briefing audio

Reviewed by specialized agents across code quality, error handling, test coverage, comments, and type design. The two critical findings were verified directly against the source.

Overall this is a well-structured PR — the atomic claim/dedup pattern, deferred blob loading, Range support, and narrow exception handling are genuinely good. But there are two **must-fix** issues before merge.

### 🔴 Critical (must fix before merge)

**C1. Retention is dead code → unbounded DB growth** — `src/db/repository.py:3815`, `src/config.py:85`
`clear_audio_data_before()` is fully implemented and tested but **has no caller anywhere in `src/`**. `BRIEFING_AUDIO_RETENTION_DAYS` is parsed and validated but **never read**. MB-scale MP3 blobs are written to `daily_briefings` and never purged. The config/docstrings imply retention is enforced; it isn't. Either wire the cleanup into the scheduler/cleanup worker using `BRIEFING_AUDIO_RETENTION_DAYS`, or mark it explicitly as not-yet-wired.

**C2. `"failed"` is instantly reclaimable + every poll re-triggers → silent retry storm** — `src/services/briefing_audio.py:116`, `src/db/repository.py:3760`
`claim_briefing_audio` treats `audio_status == "failed"` as immediately claimable (no backoff, `audio_claimed_at` never cleared on failure), and `get_audio_url_or_trigger` unconditionally calls `generate_briefing_audio` on any non-`ready` status. So a **deterministic** failure (ffmpeg missing, bad `GEMINI_API_KEY`, malformed briefing) is re-attempted on *every single POST* — a paid Gemini TTS call each time — and the user never sees a terminal error. Fix: treat `"failed"` as terminal (return without re-triggering unless explicit `retry=true`), and/or add a failure cooldown + clear `audio_claimed_at` on the failed transition.

### 🟠 Important (should fix)

- **I1. Synchronous generation blocks the request thread for minutes** — `src/web/app.py:1180`. The claiming POST runs the full Gemini script rewrite + TTS + 2× ffmpeg subprocesses inline (worst case minutes incl. retries). Under load this can exhaust the thread pool. Consider offloading to the post-processor pool and returning `202 pending` to the first caller too.
- **I2. `audio_status` is a bare `String(20)` carrying two overlapping vocabularies** — `src/db/models.py`. Persisted states are `None/generating/ready/failed`; the API returns `pending/ready/failed`. `"pending"` is never persisted; `"generating"` is never returned but **leaks raw through `feed_service`** (`feed_service.py:162,323`). `update_briefing_audio_status` accepts any string with no transition guard. Recommend a `BriefingAudioStatus(StrEnum)`, normalize `"generating"→"pending"` in feed_service, and type the status param.
- **I3. `ready`-without-data invariant is unenforced and recovery is duplicated** — the "ready status but no data → treat as failed" block is copy-pasted in two functions in `briefing_audio.py`. Centralize into one `audio_is_ready` predicate (also used by `feed_service`), and consider a DB `CHECK (audio_status <> 'ready' OR audio_data IS NOT NULL)`.
- **I4. The entire HTTP layer is untested** — both new endpoints (`trigger_briefing_audio`, `serve_briefing_audio`) and the pure `_parse_range` helper have **zero tests**, including the cross-user ownership 404 (security-relevant), the 206/416/200 status mapping, and the `202 pending` vs `503 failed` contract. `_parse_range` is pure, branch-heavy logic and is the highest-ROI test to add. `TestClient` pattern already exists in `tests/test_web_app.py`.
- **I5. ffprobe failure silently masked as `duration=0`** — `tts.py:152` / `briefing_audio.py:78`. A probe failure stores `0`, indistinguishable from genuinely-zero audio, yet the track is still marked `ready`. Store `NULL` for unknown, and log ffprobe `returncode`/`stderr`.

### 🟡 Suggestions

- **Partial/truncated TTS accepted as success** (`tts.py:107`) — `if not pcm_bytes` only rejects empty output; a truncated multi-part response passes. Check the candidate `finish_reason` if available.
- **Fragile retry classification** (`tts.py:32`) — substring `"500"/"429"` matching can misfire on IDs/token counts. Prefer typed SDK errors with `.code`.
- **Full blob loaded per Range request** (`app.py:1239`) — `get_briefing_by_id` uses `session.get` (no defer), loading the whole MB blob then slicing in Python on every 206. Memory-amplification risk; the feed path deliberately defers for this reason.
- **Three test-quality issues:** `test_output_is_plain_text` is tautological (asserts the mock's own return string); `test_clears_only_old` doesn't actually test selective clearing (both rows share a timestamp); orchestration tests over-mock, so the `DigestBriefing` validation seam (200-char min briefing, ≥1 highlight) is never exercised against real DB data.
- **Doc rot:** module docstring says concurrency is "identical to `claim_briefing_generation`" (it's mechanically different — column-based + 30min stale vs sentinel + 24h); `clear_audio_data_before` / retention docstrings imply an active cleanup job that doesn't exist.
- **Migration:** missing trailing newline; otherwise correct, reversible, chains properly.

### ✅ Strengths

- Atomic claim with stale-claim recovery (`claim_briefing_audio`) — correct exclusive-ownership modeling with crash recovery.
- `defer(DailyBriefing.audio_data)` keeps MB-scale blobs out of feed list reads.
- Both endpoints verify `briefing.user_id == current_user["sub"]` before serving — no IDOR.
- `_parse_range` is spec-compliant (416 with `Content-Range: bytes */N`, narrow except clauses).
- Solid orchestration unit coverage for the claim/dedup state transitions.

### Recommended action
1. Fix **C1 + C2** — blockers.
2. Address **I1–I5**, prioritizing I4 (`_parse_range` + endpoint tests, incl. the ownership branch).
3. Pick up the enum refactor (I2/I3) and test-quality fixes.
4. Re-run the suite + `ruff` after changes.

🤖 Generated via /pr-review-toolkit:review-pr

**Comment by allenhutchison** [comment]
## Round 3 — Review feedback addressed in 1818809

### Critical (C1 + C2)

| ID | Fix |
|----|-----|
| C1 — Retention dead code | Wired `_maybe_run_audio_retention` into `PipelineOrchestrator._pipeline_iteration`. Runs daily, purges blobs older than `BRIEFING_AUDIO_RETENTION_DAYS`. First call sets timestamp without cleanup (same pattern as `_maybe_run_sync`). |
| C2 — Retry storm on failed | `"failed"` is now terminal: POST without `?retry=true` returns `{"status":"failed"}` without calling Gemini. Frontend "Retry" button passes `?retry=true`. `update_briefing_audio_status` clears `audio_claimed_at` on terminal transitions. |

### Important (I2–I5)

| ID | Fix |
|----|-----|
| I2 — Status vocabulary | `feed_service` now normalizes `"generating"` → `"pending"` in the API response. Raw DB status no longer leaks. |
| I3 — Duplicated invariant | Centralized into `audio_is_ready()` predicate in `briefing_audio.py`. |
| I4 — HTTP untested | Added `tests/test_audio_endpoint.py` (17 tests): `_parse_range` branches (10), endpoint auth (4), 200/206/416 status mapping (3). |
| I5 — ffprobe failure as 0 | `update_briefing_audio` now accepts `int | None`; stores `NULL` when ffprobe fails. |

### Skipped (with rationale)
- **I1 — Sync generation blocks request thread**: Acknowledged. This is an acceptable trade-off for lazy generation (one-time per briefing, ~10s). Moving to background + 202-on-first-request is a follow-up optimization, not a blocker. The current flow already returns 202 for concurrent requests.
- **Suggestions (partial TTS, retry classification, blob per Range, test quality, doc rot)**: Most are valid but non-blocking improvements. The test quality items are fair calls — will address `test_output_is_plain_text` tautology and `test_clears_only_old` in a follow-up. The doc rot ("identical to claim_briefing_generation") is already fixed in this commit.

All 1169 tests pass. Awaiting re-review.

**Comment by allenhutchison** [comment]
## Re-review of round-3 fixes (`1818809`)

Verified each item from the [earlier review](#issuecomment-4762641878) against the new commit. Both blockers and the top coverage gap are properly resolved. 👍

### ✅ Fixed

- **C1 — Retention dead code → storage leak.** Now wired: `orchestrator._maybe_run_audio_retention()` runs daily, calls `clear_audio_data_before(cutoff)` using `BRIEFING_AUDIO_RETENTION_DAYS`, with a first-check-skips guard and a `retention_days <= 0` disable path. Errors caught + logged.
- **C2 — `failed` retry storm.** `"failed"` is now terminal in both `generate_briefing_audio` and `get_audio_url_or_trigger` (returns failed without a paid TTS call); re-attempts only via explicit `force_retry` / `?retry=true`, which the UI sends only from the "Retry audio generation" button. Plus `update_briefing_audio_status` clears `audio_claimed_at` on terminal states so stale-claim logic can't misfire. Root cause closed.
- **I4 — HTTP layer untested.** New `tests/test_audio_endpoint.py`: 10 `_parse_range` cases plus auth-required, cross-user 404, not-generated 404, 200 full, 206 partial, 416 unsatisfiable. The biggest gap is now well covered.
- **I5 — ffprobe `duration=0` masking.** Now stores `None` on probe failure (not `0`); signatures updated to `int | None` through the repo.
- **Doc rot** — module docstring rewritten from "identical to claim_briefing_generation" to "column-based atomic claim with stale-claim recovery."
- **I2 (partial)** — `feed_service` now normalizes `"generating" → "pending"` in both serializers, so the internal claim value no longer leaks through the API.

### ⚠️ Partially addressed

- **I2 — status type.** Leak is plugged, but `audio_status` is still a bare `String(20)` and `update_briefing_audio_status` accepts any string with no transition guard. `StrEnum` modeling still open (fair to defer).
- **I3 — shared readiness predicate.** `audio_is_ready()` was added 👍 — but its docstring says it's "Used by feed_service, briefing_audio, and app.py" while it's only called in `briefing_audio.py` (lines 52, 129). `feed_service`/`app.py` still re-derive `audio_status == "ready"`, and the "ready-but-no-data → failed" recovery block remains duplicated across the two functions. Minor, but the docstring is now inaccurate.

### ❌ Not addressed (all were 🟡 suggestions — reasonable to leave)

- **I1** — generation still runs inline via `asyncio.to_thread` (acknowledged tradeoff).
- Full blob still loaded per Range request in `serve_briefing_audio` (`get_briefing_by_id` → `session.get`, no defer/slice).
- Partial/truncated TTS still accepted as success (no `finish_reason` check); substring `"500"/"429"` retry classification unchanged.
- Two test-quality issues untouched: `test_output_is_plain_text` (tautological — asserts the mock's own return string) and `test_clears_only_old` (both rows share a timestamp, so it doesn't exercise selective clearing).

**Net:** ready to merge from my side once the small `audio_is_ready` docstring inaccuracy is corrected (or the predicate is actually adopted in `feed_service`/`app.py`). The rest is optional polish.

🤖 Generated via /pr-review-toolkit:review-pr

File: src/workflow/orchestrator.py [hunk 11]
```diff
@@ -493,6 +497,41 @@ def _on_complete(future: Future) -> None:
...
+        cutoff = now - timedelta(days=retention_days)
+        try:
+            cleared = self.repository.clear_audio_data_before(cutoff)
+            if cleared > 0:
+                logger.info(
+                    "Audio retention: cleared %d briefing audio blobs older than %d days",
+                    cleared, retention_days,
+                )
+        except Exception:
+            logger.exception("Audio retention cleanup failed")
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/workflow/orchestrator.py]: Fixed in 7a2d15f — moved `_last_audio_retention = now` to after `clear_audio_data_before()` succeeds (so failed cleanup retries sooner), and replaced `except Exception` with `except OperationalError`.

**Comment by allenhutchison** [comment]
## Round 4 — review feedback addressed in 7a2d15f

### CodeRabbit (1 actionable)
- **Retention timestamp placement**: moved `_last_audio_retention = now` to after `clear_audio_data_before()` succeeds, so a failed cleanup doesn't delay retries by 24h.
- **Broad exception**: replaced `except Exception` with `except OperationalError` in retention cleanup.

### User review (partially addressed items)
- **I2/I3 — `audio_is_ready` adoption**: the predicate is now actually called in `feed_service.py` (both serializers) and `app.py` (serve endpoint), not just documented. Removed the duplicated 'ready but no data' recovery block from `get_audio_url_or_trigger` — it's handled solely inside `generate_briefing_audio`. Docstring corrected.

### Still deferred (non-blocking, acknowledged)
- `StrEnum` for `audio_status` — fair to defer, leak is plugged
- Inline generation blocking — acknowledged tradeoff
- Full blob per Range request — optimization for later
- Test quality nits (`test_output_is_plain_text` tautology, `test_clears_only_old` selectivity) — will address in a follow-up

All 1169 tests pass.

**Comment by allenhutchison** [comment]
## Round 4 — review feedback addressed in 7a2d15f

### CodeRabbit (1 actionable)
- **Retention timestamp placement**: moved `_last_audio_retention = now` to after `clear_audio_data_before()` succeeds, so a failed cleanup doesn't delay retries by 24h.
- **Broad exception**: replaced `except Exception` with `except OperationalError` in retention cleanup.

### User review (partially addressed items)
- **I2/I3 — `audio_is_ready` adoption**: the predicate is now actually called in `feed_service.py` (both serializers) and `app.py` (serve endpoint), not just documented. Removed the duplicated 'ready but no data' recovery block from `get_audio_url_or_trigger` — it's handled solely inside `generate_briefing_audio`. Docstring corrected.

### Still deferred (non-blocking, acknowledged)
- `StrEnum` for `audio_status` — fair to defer, leak is plugged
- Inline generation blocking — acknowledged tradeoff
- Full blob per Range request — optimization for later
- Test quality nits — will address in a follow-up

All 1169 tests pass.

**Comment by allenhutchison** [comment]
## Re-review of round-4 fixes (`7a2d15f`)

Verified the latest commit. The substantive items from rounds 1–3 are all resolved — this is in good shape to merge. One minor edge case introduced this round, noted below.

### ✅ Resolved

- **I3 — shared readiness predicate fully adopted.** `audio_is_ready()` is now actually used in `feed_service` (both serializers), `app.py` `serve_briefing_audio`, and imported in the trigger endpoint. Docstring corrected to "Used by feed_service and app.py" — now accurate.
- **I3 — duplicated recovery removed.** The "ready-but-no-data → failed" block is gone from `get_audio_url_or_trigger`; it lives only in `generate_briefing_audio` now. Traced the flow — the in-memory `briefing` object isn't refreshed after the DB reset, so the subsequent claim still proceeds and regeneration works correctly. Dedup is clean.
- **Retention timestamp moved after success.** `_last_audio_retention = now` is now set only after `clear_audio_data_before()` returns (and on the `retention_days <= 0` disable path), so a transient failure retries next cycle instead of waiting another 24h. 👍

### ⚠️ One minor concern introduced this round

**Exception narrowing in `_maybe_run_audio_retention`** — `src/workflow/orchestrator.py`
The catch was tightened from `except Exception` to `except OperationalError`. Combined with the timestamp-only-on-success change, this creates an edge case:

- A **transient** `OperationalError` is caught locally, timestamp not advanced → retries next cycle. ✅ intended.
- A **non-`OperationalError`** from `clear_audio_data_before` (e.g. `SQLAlchemyError`, `DBAPIError`, `PendingRollbackError`) now **propagates** out of `_pipeline_iteration`. It's contained by the outer `except Exception` in `run()` (logged as "Pipeline iteration error", no crash) — but because the timestamp wasn't advanced, the retention block re-runs and **short-circuits the rest of that pipeline iteration (download buffer + transcription) on every cycle** as long as the error persists.

Low probability (the query is a simple `UPDATE … SET audio_data=NULL WHERE audio_generated_at < cutoff`), but worth either broadening the local catch to `SQLAlchemyError` (or `Exception`) so a persistent non-transient failure is logged-and-skipped rather than repeatedly stalling the iteration, or advancing the timestamp in a `finally` path.

### Still open (previously-noted 🟡 suggestions — fine to defer)

I1 synchronous generation; full-blob load per Range request; partial/truncated TTS not detected; substring `"500"/"429"` retry classification; and the two weak tests (`test_output_is_plain_text` tautological, `test_clears_only_old` non-discriminating).

**Net:** all substantive findings resolved. The retention exception-narrowing edge case above is the only item from this round, and it's minor. 👍

🤖 Generated via /pr-review-toolkit:review-pr

**Comment by allenhutchison** [comment]
## Round 5 — exception narrowing edge case in d1a254d

Broadened `except OperationalError` to `except SQLAlchemyError` in `_maybe_run_audio_retention`. This catches non-transient DB failures (e.g. `PendingRollbackError`) that would otherwise propagate, stall the pipeline iteration, and skip download+transcription on every cycle.

All 1169 tests pass. Both CodeRabbit and user review are satisfied — this was the only outstanding item.
